src/metabase/query_processor/pipeline.clj

```diff
@@ -0,0 +1,108 @@
+
+(defn ^:dynamic *reduce*
+  "Called by [[*run*]] (inside the `respond` callback provided by it) to reduce results of query. Reduces results, then
+  calls [[*result*]] with the reduced results."
+  [rff metadata reducible-rows]
+  (when-not (canceled?)
+    (let [[status rf-or-e] (try
+                             [::ready-to-reduce (rff metadata)]
+                             (catch Throwable e
+                               [::error (ex-info (i18n/tru "Error building query results reducing function: {0}" (ex-message e))
+                                                 {:type qp.error-type/qp, :rff rff}
+                                                 e)]))
+          [status result]  (case status
+                             ::ready-to-reduce
+                             (try
+                               [::success (transduce identity rf-or-e reducible-rows)]
+                               (catch Throwable e
+                                 [::error (ex-info (i18n/tru "Error reducing result rows: {0}" (ex-message e))
+                                                   {:type qp.error-type/qp}
+                                                   e)]))
+
+                             ::error
+                             [status rf-or-e])]
+      (case status
+        ::success (*result* result)
+        ::error   (throw result)))))
+
+(defn- interrupted-exception?
+  "If Throwable `e` is an InterruptedException or one of its causes is."
+  [e]
+  (or (instance? InterruptedException e)
+      (some-> (ex-cause e) interrupted-exception?)))
+
+(defn ^:dynamic *run*
+  "Function for running the query. Calls [[*execute*]], then [[*reduce*]] on the results."
+  [query rff]
+  (when-not (canceled?)
+    (letfn [(respond [metadata reducible-rows]
+              (*reduce* rff metadata reducible-rows))]
+      (try
+        (*execute* driver/*driver* query respond)
+        (catch Throwable e
+          ;; rethrow e if it's not an InterruptedException, we're not interested in it.
+          (when-not (interrupted-exception? e)
+            (throw e))
+          ;; ok, at this point we know it's an InterruptedException.
+          (log/tracef e "Caught InterruptedException when executing query, this means the query was canceled. Ignoring exception.")
+          ;; just to be extra safe and sure that the canceled chan has gotten a message. It's a promise channel so
+          ;; duplicate messages don't matter
+          (some-> *canceled-chan* (a/>!! ::cancel))
+          ::cancel)))))
+
+(def ^:dynamic ^Long *query-timeout-ms*
+  "Maximum amount of time query is allowed to run, in ms."
+  ;; I don't know if these numbers make sense, but my thinking is we want to enable (somewhat) long-running queries on
+  ;; prod but for test and dev purposes we want to fail faster because it usually means I broke something in the QP
+  ;; code
+  (u/minutes->ms
+   (if config/is-prod?
+     20
+     3)))
```

Comment:
This seems to be going in the opposite direction we took (or wanted to take) with the API stuff: there we said we would prefer fewer dynamic vars, didn't we?

Maybe we can implement a function that we could call when debugging that would give the values of the "context" vars and also say if they have been redefined.